transport/grpc.go

```diff
@@ -0,0 +1,184 @@
+	return newGRPC(options)
+}
+
+type grpcTransport struct {
+	Transport       transport.Transport
+	Outbound        transport.UnaryOutbound
+	Caller          string
+	Encoding        string
+	RoutingKey      string
+	RoutingDelegate string
+	tracer          opentracing.Tracer
+}
+
+func newGRPC(options GRPCOptions) (*grpcTransport, error) {
+	if len(options.Addresses) == 0 {
+		return nil, errGRPCNoAddresses
+	}
+	if options.Tracer == nil {
+		return nil, errGRPCNoTracer
+	}
+	if options.Caller == "" {
+		return nil, errGRPCNoCaller
+	}
+	if options.Encoding == "" {
+		return nil, errGRPCNoEncoding
+	}
+
+	identifiers := make([]apipeer.Identifier, len(options.Addresses))
+	for i, address := range options.Addresses {
+		identifiers[i] = hostport.Identify(address)
+	}
+	transport := grpc.NewTransport(grpc.Tracer(options.Tracer))
+	outbound := transport.NewOutbound(peer.Bind(roundrobin.New(transport), peer.BindPeers(identifiers)))
+	if err := transport.Start(); err != nil {
+		return nil, err
+	}
+	if err := outbound.Start(); err != nil {
+		_ = transport.Stop()
+		return nil, err
+	}
+
+	return &grpcTransport{
+		Transport:       transport,
+		Outbound:        outbound,
+		Caller:          options.Caller,
+		Encoding:        options.Encoding,
+		RoutingKey:      options.RoutingKey,
+		RoutingDelegate: options.RoutingDelegate,
+		tracer:          options.Tracer,
+	}, nil
+}
+
+func (t *grpcTransport) Tracer() opentracing.Tracer {
+	return t.tracer
+}
+
+func (t *grpcTransport) Protocol() Protocol {
+	return GRPC
+}
+
+func (t *grpcTransport) Call(ctx context.Context, request *Request) (*Response, error) {
+	if request.TargetService == "" {
+		return nil, errGRPCNoService
+	}
+	if request.Method == "" {
+		return nil, errGRPCNoProcedure
+	}
+
+	ctx, cancel := requestContextWithTimeout(ctx, request)
+	defer cancel()
+	transportResponse, err := t.Outbound.Call(ctx, t.requestToTransportRequest(request))
+	if err != nil {
+		return nil, err
+	}
+	return transportResponseToResponse(transportResponse)
+}
+
+func (t *grpcTransport) Close() error {
+	return multierr.Combine(t.Transport.Stop(), t.Outbound.Stop())
+}
+
+func (t *grpcTransport) requestToTransportRequest(request *Request) *transport.Request {
+	return &transport.Request{
+		Caller:          t.Caller,
+		Service:         request.TargetService,
+		Encoding:        transport.Encoding(t.Encoding),
+		Procedure:       request.Method,
+		Headers:         transport.HeadersFromMap(request.Headers),
+		ShardKey:        request.ShardKey,
+		RoutingKey:      t.RoutingKey,
+		RoutingDelegate: t.RoutingDelegate,
+		Body:            bytes.NewReader(request.Body),
+	}
+}
+
+func requestContextWithTimeout(ctx context.Context, request *Request) (context.Context, context.CancelFunc) {
+	if _, ok := ctx.Deadline(); ok {
+		return ctx, noop
+	}
+	timeout := time.Second
+	if request.Timeout > 0 {
+		timeout = request.Timeout
+	}
+	return context.WithTimeout(ctx, timeout)
```

Comment:
do we need this at all? i think that the caller should always pass a context with the timeout, they just include the value in the `Request` so that you can pass it through headers for propagation if needed.

Reply:
I don't think the caller always does this, and this is for safety/to take into account the impls in the other transports.